auth_totp/controllers/main.py

```diff
@@ -0,0 +1,124 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 LasLabs Inc.
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from datetime import datetime, timedelta
+import json
+from werkzeug.contrib.securecookie import SecureCookie
+from openerp import http, registry, SUPERUSER_ID
+from openerp.api import Environment
+from openerp.http import Response, request
+from openerp.addons.web.controllers.main import Home
+from ..exceptions import MfaTokenInvalidError, MfaTokenExpiredError
+
+
+class JsonSecureCookie(SecureCookie):
+    serialization_method = json
+
+
+class AuthTotp(Home):
+    @http.route()
+    def web_login(self, *args, **kwargs):
+        response = super(AuthTotp, self).web_login(*args, **kwargs)
+
+        if not request.params.get('login_success'):
+            return response
+
+        user = request.env['res.users'].sudo().browse(request.uid)
+        if not user.mfa_enabled:
+            return response
+
+        cookie_key = 'trusted_devices_%s' % user.id
+        device_cookie = request.httprequest.cookies.get(cookie_key)
+        if device_cookie:
+            config_model = request.env['ir.config_parameter']
+            secret = config_model.sudo().get_param('database.secret')
+            device_cookie = JsonSecureCookie.unserialize(device_cookie, secret)
+            if device_cookie.get('device_id') in user.trusted_device_ids.ids:
+                return response
+
+        user.generate_mfa_login_token()
+        request.session.logout(keep_db=True)
+        return http.local_redirect(
+            '/auth_totp/login',
+            query={
+                'mfa_login_token': user.mfa_login_token,
+                'redirect': request.params.get('redirect'),
+            },
+            keep_hash=True,
+        )
+
+    @http.route('/auth_totp/login', type='http', auth='none', methods=['GET'])
+    def mfa_login_form(self, *args, **kwargs):
+        return request.render('auth_totp.mfa_login', qcontext=request.params)
+
+    @http.route('/auth_totp/login', type='http', auth='none', methods=['POST'])
+    def mfa_login(self, *args, **kwargs):
+        token = request.params.get('mfa_login_token')
+        try:
+            user = request.env['res.users'].user_from_mfa_login_token(token)
+        except (MfaTokenInvalidError, MfaTokenExpiredError) as exception:
+            if isinstance(exception, MfaTokenInvalidError):
```

Comment:
@moylop260 - Unfortunately, that option is off the table because these inherit from `AccessDenied`, whose constructor doesn't accept a message: https://github.com/odoo/odoo/blob/9.0/openerp/exceptions.py#L52